contents/blog/temporal-exports.md

```diff
@@ -0,0 +1,118 @@
+
+Our export system provides the ability to maintain ownership of all of your data in PostHog and move it where you like. Unfortunately, it didn’t scale as well as we wanted. Luckily for you, we recently launched our [newly upgraded exports system](/docs/cdp/batch-exports), built using Temporal. This post covers why and how we made the change to Temporal.
+
+## The trouble with exports
+
+Although exports worked well most of the time, they were causing trouble for some customers. They were failing, erroring, and dropping data. Fixing these issues was taking up a significant amount of the pipeline team’s time. 
+
+Although stateless plugins (like GeoIP) worked well with our [plugin service structure](/blog/how-we-built-an-app-server), **stateful** plugins (like exports) didn’t. This is because they depended on external resources that can degrade or fail. Dealing with this required timeout and retry logic that was difficult  to build into the existing structure.
+
+Our old export system scaled as far as possible. It worked well for smaller users, but as the export requirements grew, it started to show issues. Chiefly among them included:
+
+### Unreliable (not retrying)
+
+The biggest issue was not handling errors gracefully, which made the entire system unreliable. This is largely due to the statefulness of exports. A failure meant they lost their state which was unrecoverable.
+
+For example, there are transient errors fixed by retrying exports (such as temporary network disconnects), but this wasn’t built into the structure of exports. They needed to be manually retried which was unsustainable. Other exports allowed resets, but they didn’t make progress once reset. This clogged the bucket and pipeline causing resets to do more harm than good.
+
+Another big issue was needing to run entire batches again if they failed. It wasn’t possible to restart from where we left off. Exports couldn’t pause or run simultaneously either. This put a lot of stress on the batch systems. 
+
+Altogether, this created an inconsistent, unreliable export system.
+
+### Opaque
+
+On top of unreliability, we didn’t have a clear picture of what was happening in the export system. Throughout our codebase, we had multiple different work queues like Celery, Kafka, and Graphile each running separately in different contexts.
+
+We didn’t have logs or reports. For example, we didn’t know how much data was being exported or when duplicate runs were happening. When issues happened, we relied on users to tell us. This led to a lot of firefighting and often manual restarts of exports.
+
+### Expensive
+
+The final nail was that exports were expensive. This wasn’t because we were charging a lot for exports, we weren’t charging anything. It was because they were expensive on the destination side. 
```

Comment:
```suggestion
The final nail was that exports were expensive. This wasn’t because we were charging a lot for exports – in fact, we weren’t charging anything. It was because they were expensive on the destination side. 
```